### PR TITLE
[Merged by Bors] - miner: use atxsdata to generate activeset and persist it

### PR DESCRIPTION
**Pull request by dshulyak**
related: https://github.com/spacemeshos/go-spacemesh/issues/5192 https://github.com/spacemeshos/go-spacemesh/issues/5106

this change accomplishes three things:
- removes lru cache from miner module, after this change it won't be on paths sensitive to timings and can be reduced
- allows to prepare activeset in advance, before epoch starts
- persists prepared activeset, so that node doesn't have to redo the work after restart

in order to prepare activeset we don't have to wait for the first layer of the epoch, all the data required for activeset
must be downloaded as soon as two hours prior to epoch start, but in configuration i set preparation to start 30m before the epoch start.

also this process becomes more and more expensive with the number of atxs growth, and there is a benefit of persisting activeset locally.

### CODE REVIEW

File: miner/active_set_generator.go [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+// ensure tries to generate active set for the target epoch in configured number of tries,
+// each try will be repeated after configured retry interval.
+func (a *activeSetGenerator) ensure(ctx context.Context, target types.EpochID) {
+	var err error
+	for try := 0; try < a.cfg.activeSet.Tries; try++ {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(a.cfg.activeSet.RetryInterval):
+		}
```
Comment (poszu) [inline miner/active_set_generator.go]: How about moving the wait below `a.log.Debug("failed to prepare active set"` so it doesn't start with waiting?

File: miner/active_set_generator.go [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+			zap.Duration("network delay", networkDelay),
+		)
+		result, err := activeSetFromGrades(p.db, target, epochStart, networkDelay)
+		if err != nil {
+			return id, 0, nil, err
+		}
+		if result.Total == 0 {
+			return id, 0, nil, fmt.Errorf("empty active set")
+		}
+		if result.Total > 0 && len(result.Set)*100/result.Total > p.cfg.goodAtxPercent {
```
Comment (poszu) [inline miner/active_set_generator.go]: The `result.Total > 0` is already satisfied by the check above.
```suggestion
		if result.Total == 0 {
			return id, 0, nil, fmt.Errorf("empty active set")
		}
		if len(result.Set)*100/result.Total > p.cfg.goodAtxPercent {
```

File: miner/active_set_generator.go [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+	)
+	if err := atxs.IterateForGrading(db, target-1, func(id types.ATXID, atxtime, prooftime int64, weight uint64) bool {
+		total++
+		if gradeAtx(epochStart, networkDelay, atxtime, prooftime) == good {
+			set = append(set, id)
+			setWeight += weight
+		}
+		return true
+	}); err != nil {
+		return gradedActiveSet{}, fmt.Errorf("failed to iterate atxs that target epoch %v: %v", target, err)
```
Comment (poszu) [inline miner/active_set_generator.go]:
```suggestion
		return gradedActiveSet{}, fmt.Errorf("failed to iterate atxs that target epoch %v: %w", target, err)
```

File: miner/proposal_builder.go [hunk 4]
```diff
@@ -291,38 +312,66 @@ func (pb *ProposalBuilder) Register(signer *signing.EdSigner) {
...
+					} else {
+						pb.logger.With().Warning("failed to build proposal",
+							log.Context(ctx), log.Uint32("lid", current.Uint32()), log.Err(err),
+						)
+					}
 				}
 			}
 		}
+	})
+	return eg.Wait()
```
Comment (poszu) [inline miner/proposal_builder.go]: What is the point of running it on a goroutine and block here? How about dropping `eg.Go()` and running the code synchronously like it was before?

File: sql/atxs/atxs.go [hunk 5]
```diff
@@ -539,3 +539,30 @@ func CountAtxsByOps(db sql.Executor, operations builder.Operations) (count uint3
...
+
+// IterateForGrading selects every atx from publish epoch and joins identities to load malfeasence proofs if they exist.
+func IterateForGrading(
+	db sql.Executor,
+	epoch types.EpochID,
+	fn func(id types.ATXID, atxtime, prooftime int64, weight uint64) bool,
+) error {
+	if _, err := db.Exec(`
+		select atxs.id as id, atxs.received as atxtime,
+			   identities.received as prooftime, effective_num_units, tick_count from atxs
```
Comment (poszu) [inline sql/atxs/atxs.go]: Does `as foo` have any effect here? The values are referred to by col index anyway.
```suggestion
		select atxs.id, atxs.received, identities.received, effective_num_units, tick_count
		from atxs
```

File: miner/active_set_generator.go (hunk 1 above)
Comment (dshulyak) [reply miner/active_set_generator.go]: makes sense

File: sql/atxs/atxs.go (hunk 5 above)
Comment (dshulyak) [reply sql/atxs/atxs.go]: it doesn't, i added it for my clarity, can drop

**Review by poszu — APPROVED** [review]

**Comment by dshulyak** [comment]
i will merge it after atx sync improvements

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors cancel

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**